autoload/neomake/signs.vim

```diff
@@ -0,0 +1,188 @@
+    let s:sign_queue[a:entry.bufnr] = get(s:sign_queue, a:entry.bufnr, {})
+    let existing = get(s:sign_queue[a:entry.bufnr], a:entry.lnum, {})
+    if empty(existing) || a:entry.type ==# 'E' && existing.type !=# 'E'
+        let s:sign_queue[a:entry.bufnr][a:entry.lnum] = a:entry
+    endif
+endfunction
+
+function! neomake#signs#PlaceSign(existing_signs, entry) abort
+    let type = a:entry.type ==# 'E' ? 'neomake_err' : 'neomake_warn'
+
+    let a:existing_signs.by_line[a:entry.lnum] = get(a:existing_signs.by_line,
+                                                   \ a:entry.lnum, [])
+    let s:placed_signs[a:entry.bufnr] = get(s:placed_signs, a:entry.bufnr, {})
+    if !has_key(s:placed_signs[a:entry.bufnr], a:entry.lnum)
+        let sign_id = a:existing_signs.max_id + 1
+        let a:existing_signs.max_id = sign_id
+        let cmd = 'sign place '.sign_id.' line='.a:entry.lnum.
+                                      \ ' name='.type.
+                                      \ ' buffer='.a:entry.bufnr
+        let s:placed_signs[a:entry.bufnr][a:entry.lnum] = sign_id
+    elseif type ==# 'neomake_err'
+        " Upgrade this sign to an error
+        let sign_id = s:placed_signs[a:entry.bufnr][a:entry.lnum]
+        let cmd =  'sign place '.sign_id.' name='.type.' buffer='.a:entry.bufnr
+    else
+        let cmd = ''
+    endif
+
+    if len(cmd)
+        call neomake#utils#DebugMessage('Placing sign: '.cmd)
+        exe cmd
+    endif
+
+    " Replace all existing signs for this line, so that ours appear on top
+    for existing in get(a:existing_signs.by_line, a:entry.lnum, [])
+        if existing.name !~# 'neomake_'
+            exe 'sign unplace '.existing.id.' buffer='.a:entry.bufnr
+            exe 'sign place '.existing.id.' line='.existing.line.
+                                        \ ' name='.existing.name.
+                                        \ ' buffer='.a:entry.bufnr
+        endif
+    endfor
+endfunction
+
+function! neomake#signs#CleanOldSigns() abort
+    call neomake#utils#DebugObject('Cleaning old signs:', s:last_placed_signs)
+    for buf in keys(s:last_placed_signs)
+        for ln in keys(s:last_placed_signs[buf])
+            let cmd = 'sign unplace '.s:last_placed_signs[buf][ln]
+            call neomake#utils#DebugMessage('Unplacing sign: '.cmd)
+            exe cmd
+        endfor
+    endfor
+    let s:last_placed_signs = {}
+endfunction
+
+function! neomake#signs#PlaceVisibleSigns() abort
+    let buf = bufnr('%')
+    if !has_key(s:sign_queue, buf)
+        return
+    endif
+    let topline = line('w0')
+    let botline = line('w$')
+    for ln in range(topline, botline)
+        if has_key(s:sign_queue[buf], ln)
+            if !exists('l:signs')
+                let l:signs = neomake#signs#GetSignsInBuffer(buf)
+            endif
+            call neomake#signs#PlaceSign(l:signs, s:sign_queue[buf][ln])
+            unlet s:sign_queue[buf][ln]
+        endif
+    endfor
```

Comment:
if `s:sign_queue[buf]` is empty at this point, maybe the key `buf` could be removed from `s:sign_queue` so next calls to this function return faster.